### PR TITLE
fix(layout): measured band (categorical) x-axis label layout — wrap/rotate, never drop a bar [#387]

### LINKED ISSUE #387: layout: band/categorical x labels collide — need wrap/rotate, not omit, for low-cardinality bars
## Summary

Band (categorical) x-axis labels on column/bar charts **collide with neighbouring labels and with the axis title** when category strings are long. Temporal x-axes already get measured, calendar-aware guide plans that avoid collisions. Categorical axes do not have an equivalent.

Worse: the only band degradations today are **thinning** (drop every *n*th label) and **ellipsis truncation** driven by the **last** label’s right-edge overhang — not by adjacent-label overlap. For bar charts, silently omitting category labels is not acceptable: every bar needs a label.

## Reproduction

```svelte
<script>
  import { GGPlot } from "@ggsvelte/svelte";

  const rows = [
    { category: "Resolución", count: 9000 },
    { category: "Corrección (errores o erratas)", count: 3200 },
    { category: "Sentencia", count: 2800 },
    { category: "Orden", count: 2500 },
    { category: "Otro", count: 900 },
  ];
</script>

<GGPlot
  height={300}
  width={480}
  inspect
  spec={{
    data: { values: rows },
    layers: [
      { geom: "col", aes: { x: { field: "category" }, y: { field: "count" } } },
    ],
  }}
/>
```

1. Render at a typical dashboard width (~400–600px).
2. Observe:
   - long labels (e.g. `Corrección (errores o erratas)`) overflow their band and **overlap** the labels of adjacent categories
   - labels also crowd / collide with the x-axis title (`category`)
3. Contrast with a temporal line chart of similar width: years/dates are spaced and labelled cleanly via the temporal guide planner.

Reproduced against `@ggsvelte/svelte@0.5.1`.

## Why this happens (code path)

`packages/core/src/layout/layout.ts`:

- Explicitly: **“Label rotation is intentionally NOT implemented (skipped per spike scope).”**
- Band ticks support `labelEvery` thinning and single-line ellipsis truncation only.
- X-axis margin logic budgets **one label line** of height (`labelH + bottomFixed`) and only reacts when the **last** centred label overhangs the right margin cap — it does **not** measure pairwise neighbour collisions across the band.
- Temporal axes go through `planTemporalAxis` / `guidePlans` (measured interval selection, contextual labels). Band axes have no parallel “measured categorical guide plan”.

So long categorical labels on a few bars: `labelEvery` stays 1 (cannot thin without hiding a category that still has a bar), truncation often does not fire (right-edge check is about the last label half-width, not mid-axis collisions), and everything is drawn on one horizontal line → collisions.

## Expected behaviour

For **low-to-moderate cardinality band axes** (typical `geom_col` / `geom_bar` with a handful of categories):

1. **Never drop a category label** as the primary collision strategy. Thinning is fine for high-cardinality continuous-like band axes; it is wrong when each tick is a named bar the user must read.
2. Prefer, in order of cheapness:
   1. **Word-boundary multi-line wrap** within the band width (e.g. `Corrección` / `(errores` / `o erratas)` on separate lines — or a smarter break on spaces / punctuation).
   2. **Rotate labels** (e.g. −30° / −45° / −90°) when wrap still collides or when labels are single long tokens.
   3. **Increase bottom margin** to fit the wrapped/rotated labels **and** keep clearance from the axis title (the title must not sit under multi-line labels).
   4. Ellipsis truncation only as a last resort, with full label available via inspect / `title` / `fullLabel` (temporal already exposes `fullLabel`).
3. Detect **adjacent-label overlap** (not only last-label right overhang) before declaring the axis fine.
4. Coordinate with axis-title layout so multi-line / rotated ticks do not collide with the x-axis title.

`coord_flip` remains a good author choice for very long labels (and is already documented in the skill recipe table), but the upright default should not require it to be readable.

## Suggested direction

- Add a measured **band guide plan** analogous to temporal `guidePlans`:
  - input: category strings, band widths, font metrics, available bottom margin cap, axis title reserve
  - output: per-tick layout mode (`single-line` | `wrapped` | `rotated`) + final margin contribution + degradations
- Cheap wrap heuristic first: split on whitespace / punctuation when `measureWidth(label) > bandWidth * factor`; re-measure multi-line block height; grow bottom margin.
- If still colliding or single-token overflow: try rotation and re-measure projected bounding box.
- Keep `labelEvery` thinning only for high-cardinality band axes (or when the author opts in); default for bar-like low cardinality should be “label every category”.
- VR coverage: long Spanish/English multi-word category labels at 480px width must not overlap neighbours or the axis title.

## Related

- Temporal axis planning already solves a harder version of “labels must stay legible without collisions” — categorical band axes need the same care with different constraints (cannot omit named categories).

### PR DESCRIPTION
**Pull request by ljodea**
Fixes #387.

## Problem
Categorical (band) x-axis labels on `geom_col`/`geom_bar` collided with each other and the axis title when category strings are long (routine in Spanish/German). Temporal axes already had a measured, collision-aware guide planner; band axes had none — their only degradations were `labelEvery` thinning (wrong for bars: every bar needs its label) and last-label right-overhang truncation (never detects pairwise neighbour overlap).

Verified against ggplot2 4.0.3 (R spike): ggplot2 does nothing automatic for discrete axes and treats `character`/`factor` identically for collision (factor only changes level order). This is a deliberate beyond-ggplot2 enhancement, exactly like our temporal planner.

## What this does
Adds `planBandAxis`, the discrete analogue of `planTemporalAxis`. On the **horizontal** band x-axis it escalates label layout *within the axis band*, honoring the author's categorical→x mapping (never auto-flip):

`single-line → wrap (≤2 lines) → rotate −45° → rotate −90° → truncate + margin-overflow`, with high-cardinality thinning.

- **Never drops a named low-cardinality bar.** `labelEvery` thinning is kept only as the high-cardinality (≥12) final rung.
- **Never auto-flips.** When rotation still can't fit, the truncation is loud (`band-label-margin-overflow`) and the full text stays on the tick `<title>` / `fullLabel`; the diagnostic `fixes[]` surfaces `coord_flip` (`.coordFlip()` / `{coord:{type:"flip"}}`) as the recommended horizontal-bar alternative.
- **Orientation-gated (G1):** vertical band (native Y, or categorical-on-Y after `coord_flip`) keeps the legacy width-cap + thin/truncate path untouched — no rotate/tspan on Y.
- **Two-pass safe (G3):** the label mode is inherited across the two layout passes (escalate-only), so margins stay A→B stable.
- **Agents see the heuristic:** `band-labels-wrapped` / `band-labels-rotated` advisories (`{code,path,chosen,howToOverride}`) are emitted from the resolved plan mode.

## Renderer
Emits `<tspan>` lines for wrapped labels and a `rotate(...)` transform for rotated ones; the single-line output is byte-identical (existing snapshots unchanged). The x-axis title y is derived from the measured label-band height so it clears multi-line / rotated labels instead of the fixed offset.

## Tests
- `band-guide.test.ts` — planner unit (ladder, escalate-only, determinism, edge cases).
- `layout.test.ts` — orientation-gated integration + monotonic-mode convergence + Y-legacy preserved.
- `render-band-labels.test.ts` — tspan/rotate SVG + single-line unchanged.
- `pipeline-band.test.ts` — end-to-end advisories + `coord_flip` diagnostic fix.
Full `bun run test` green; typecheck / oxlint / type-aware / knip / fmt clean; docs route+search projections regenerated for the new diagnostic codes.

## Reviewed
Design / DevEx / Eng plan reviews plus an adversarial `grok` plan review (which flagged this direction "needs rework"); all five blockers were folded in before implementation (margin seam before the cap; orientation gate; `AxisGuidePlan` mode/angle/height + `previousMode`; flip-aware band context; contract-test handling).

## Deferred (follow-ups)
- #406 — VR coverage for long i18n labels at 480px (baseline must be generated in the pinned Playwright container).
- #407 — author-override `guide` field (angle/wrap/mode); `coordFlip()` is the escape hatch today, and `planBandAxis` already has the `config?` seam.
- Docs guide-prose inversion ("labels wrap/rotate automatically; use coord_flip when you *want* horizontal bars") rides with #407.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/core/src/layout/band-guide.ts [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+  let labelEvery = 1;
+  let overlap = false;
+  let marginOverflow = false;
+  let orthoBudgetForLabel = orthoCap;
+
+  // Along-axis: do adjacent rotated labels still collide?
+  if (alongFootprint + gap > bandBudget) {
+    if (n >= BAND_THIN_MIN_CATEGORIES) {
+      // High cardinality: thin (never for a handful of named bars).
+      while (bandBudget * labelEvery < alongFootprint + gap && labelEvery * 2 < entries.length) {
```
[the remark this answers is not in the document]
Comment (ljodea) [reply packages/core/src/layout/band-guide.ts]: Fixed in 0412577. The rotated branch now runs `neighbourOverlap` on the **actual displayed tick positions** (`entries[i].center`), not `extentPx / categoryCount` — so a sparse break subset that is hundreds of px apart is never wrongly rotated/thinned and never hides an authored break. Regression test: `band-guide.test.ts` › "does not thin far-apart break labels even when the full domain is huge" (100-category domain, breaks at 0/50/99, asserts `labelEvery === 1`).

File: packages/core/src/pipeline/panel-layout-placements-single.ts [hunk 2]
```diff
@@ -12,11 +12,14 @@ export function placeSinglePanelFromChrome(
...
   const { h, v } = chrome.displayScales(0);
   const temporal = chrome.displayTemporal(0);
+  const band = chrome.displayBand(0);
   return placeSinglePanel({
     h,
     v,
     ...(temporal.h !== undefined && { hTemporal: temporal.h }),
     ...(temporal.v !== undefined && { vTemporal: temporal.v }),
+    ...(band.h !== undefined && { hBand: band.h }),
+    ...(band.v !== undefined && { vBand: band.v }),
```
[the remark this answers is not in the document]
Comment (ljodea) [reply packages/core/src/pipeline/panel-layout-placements-single.ts:22]: Fixed in 0412577. `displayBand` (flip-aware) is now threaded through the facet layout paths — the shared-margins pass, `placeOneFacetPanel`/`placeFacetPanelFromTicks`, and `mapFacetPanelPlacements` — so faceted bar charts receive `hBand` and get the same wrap/rotate planning, advisories, and reserved label band as single panels. Regression test: `pipeline-band.test.ts` › "plans band labels in faceted charts too" (facet cols, asserts a `band-labels-*` advisory is emitted).

File: packages/core/src/layout/layout.ts [hunk 3]
```diff
@@ -498,9 +621,18 @@ export function layoutPass(margins: Margins, input: LayoutInput, theme: LayoutTh
...
-  const rawBottom = x.empty ? minMargins.bottom : Math.max(minMargins.bottom, labelH + bottomFixed);
-  const rawRight = x.empty ? minMargins.right : Math.max(minMargins.right, lastXLabelW);
+  const rawBottom = x.empty
+    ? minMargins.bottom
+    : Math.max(minMargins.bottom, (x.bandLabelBandHeight ?? labelH) + bottomFixed);
+  const rawRight = x.empty
+    ? minMargins.right
+    : bandPlanned
+      ? Math.max(minMargins.right, x.bandAlongOverhang ?? 0)
+      : Math.max(minMargins.right, lastXLabelW);
```
[the remark this answers is not in the document]
Comment (ljodea) [reply packages/core/src/layout/layout.ts:654]: Fixed in 0412577. Single-line band axes now run `capEndOverhang`, which truncates an over-wide end label (ellipsis) using its **actual** end position, emits `band-label-margin-overflow` (full text stays on the tick `<title>`/`fullLabel`), and clamps the along-axis overhang to the margin cap so it can't be clipped at the SVG edge. Regression test: `band-guide.test.ts` › "truncates a lone over-wide label instead of clipping it at the panel edge".

**Comment by ljodea** [comment]
@codex review

Addressed all three P2 findings in 0412577:
- **Sparse-break thinning** → rotated overlap now measured at actual displayed tick positions.
- **Facet wiring** → `displayBand` threaded through the facet layout paths.
- **End-label cap** → single-line over-wide end labels truncate + emit `band-label-margin-overflow` and clamp overhang.

Please re-review the new head, especially the facet threading and `capEndOverhang`.

File: packages/core/src/layout/band-guide.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ljodea) [reply packages/core/src/layout/band-guide.ts]: **Fixed** in 0412577 — valid finding.

The rotated branch now measures neighbour overlap at the **actual displayed tick positions** (`rotatedOverlaps` → `neighbourOverlap` on filtered entries' centers), not `extentPx / categoryCount`. A sparse explicit-break subset that is hundreds of px apart is never wrongly rotated/thinned and never hides an authored break.

Regression: `planBandAxis: sparse explicit breaks (Codex P2)` in `packages/core/tests/band-guide.test.ts`.

File: packages/core/src/pipeline/panel-layout-placements-single.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ljodea) [reply packages/core/src/pipeline/panel-layout-placements-single.ts:22]: **Fixed** in 0412577 — valid finding.

`displayBand` is now threaded through the facet layout path (`panel-layout-facet-margins-pass`, `panel-layout-facet-map`, `panel-layout-facet-place-ticks`), so faceted bar charts get the same wrap/rotate planning, advisories, and reserved band as single panels.

Regression: `plans band labels in faceted charts too (Codex P2)` in `packages/core/tests/pipeline-band.test.ts`.

File: packages/core/src/layout/layout.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ljodea) [reply packages/core/src/layout/layout.ts:654]: **Fixed** in 0412577 — valid finding.

`capEndOverhang` now truncates an over-wide end label on single-line band axes (with `band-label-margin-overflow` + full text retained on the tick title) and clamps the along-axis overhang to the margin cap, instead of clipping the label at the SVG edge.

Regression: `planBandAxis: end-label margin cap (Codex P2)` in `packages/core/tests/band-guide.test.ts`.

File: packages/core/src/layout/layout.ts [hunk 4]
```diff
@@ -498,9 +621,18 @@ export function layoutPass(margins: Margins, input: LayoutInput, theme: LayoutTh
...
   const rawLeft = y.empty ? minMargins.left : Math.max(minMargins.left, yLabelW + leftFixed);
-  const rawBottom = x.empty ? minMargins.bottom : Math.max(minMargins.bottom, labelH + bottomFixed);
-  const rawRight = x.empty ? minMargins.right : Math.max(minMargins.right, lastXLabelW);
+  const rawBottom = x.empty
+    ? minMargins.bottom
+    : Math.max(minMargins.bottom, (x.bandLabelBandHeight ?? labelH) + bottomFixed);
+  const rawRight = x.empty
+    ? minMargins.right
+    : bandPlanned
+      ? Math.max(minMargins.right, x.bandAlongOverhang ?? 0)
```
[the remark this answers is not in the document]
Comment (ljodea) [reply packages/core/src/layout/layout.ts:653]: **Fixed** in 63ffc09 — valid P2.

`capEndOverhang` now tracks left and right overhang separately (`half - center` vs `half - (extentPx - center)`), and `layoutPass` folds the band left overhang into `rawLeft` (not just `rawRight`). A reversed/single/wide leftmost band label that extends past x=0 is now reserved in the left margin. Applied across all three modes (single-line, wrapped, rotated).

Regression: `band-guide.test.ts` › "reserves LEFT overhang for a centered lone label (both sides), not just right".

File: packages/core/src/layout/band-guide.ts [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+        };
+      }
+    }
+  }
+
+  // --- rotated (−45 then −90) ---
+  const maxWidth = Math.max(...entries.map((e) => e.width));
+  const along = (angle: number) => {
+    const a = Math.abs(angle) * RAD;
+    return maxWidth * Math.cos(a) + lineHeight * Math.sin(a);
```
[the remark this answers is not in the document]
Comment (ljodea) [reply packages/core/src/layout/band-guide.ts]: **Fixed** in 63ffc09 — valid P2.

The rotated overlap and orthogonal-margin checks now use each label’s **own** footprint (`alongOf(e.width, angle)` / `orthoOf(width, angle)`) rather than a single uniform `maxWidth`, and the orthogonal decision recomputes the max width from the labels that survive thinning (`labeledMaxWidth(labelEvery)`). A wide label on a hidden tick no longer inflates the overlap check nor emits a `band-label-margin-overflow` for text that will not render.

Regression: `band-guide.test.ts` › "does not emit margin-overflow for a wide label that thinning hides".

File: packages/core/src/layout/band-guide.ts [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+
+  const degraded: string[] = [];
+  let labelEvery = 1;
+  let overlap = false;
+  let marginOverflow = false;
+  let orthoBudgetForLabel = orthoCap;
+
+  // Along-axis: do adjacent rotated labels still collide at their real positions?
+  if (rotatedOverlaps(angle, 1)) {
+    if (n >= BAND_THIN_MIN_CATEGORIES) {
```
[the remark this answers is not in the document]
Comment (ljodea) [reply packages/core/src/layout/band-guide.ts]: **Fixed** in 63ffc09 — valid P2.

Thinning is now gated on `entries.length` (the displayed/authored tick count), not `n` (the full category count). A small set of close-together authored breaks in a huge domain now reports overlap and keeps every break instead of thinning one away.

Regression: `band-guide.test.ts` › "never thins a handful of close-together authored breaks in a huge domain" (breaks at adjacent indices 0/1/2 of a 100-category domain → `labelEvery === 1`, `overlap === true`).

File: packages/core/src/pipeline/panel-layout-facet-place-ticks.ts [hunk 7]
```diff
@@ -39,8 +41,8 @@ export function placeFacetPanelFromTicks(input: FacetPanelTicksInput): PanelPlac
     {
       width: panelW + mMax.left + mMax.right,
       height: panelH + mMax.top + mMax.bottom,
-      x: layoutDomain(h, hBreaks, hTemporal),
-      y: layoutDomain(v, vBreaks, vTemporal),
+      x: layoutDomain(h, hBreaks, hTemporal, hBand),
```
[the remark this answers is not in the document]
Comment (ljodea) [reply packages/core/src/pipeline/panel-layout-facet-place-ticks.ts:44]: **Fixed** in 63ffc09 — valid P2.

`computeFacetGridGeometry` now runs a second shared-margin pass at the **final** derived panel size (`cells.panelW`/`panelH`) when any panel carries a band guide plan, takes the element-wise max of the two margin sets, and re-derives the cell geometry from it. The escalated wrapped/rotated band measured at the narrower real width is therefore reserved, so bottom-row labels / the x title cannot draw into adjacent chrome. Gated to band facets, so temporal/continuous facets keep their exact single-pass geometry (no regression), and the margin can only grow (element-wise max).

Regression: `pipeline-band.test.ts` › "re-plans the narrower final facet panel so the band margin is reserved" (two-column facet at 420px still escalates to a rotated advisory).

File: packages/core/src/pipeline/assemble-scene-panels.ts [hunk 8]
```diff
@@ -155,7 +155,27 @@ export function assembleScenePanels(input: {
 
   const firstX = scenePanels.find((p) => p.axisX !== null);
   const firstY = scenePanels.find((p) => p.axisY !== null);
-  const xAxis: SceneAxis = { ticks: firstX?.axisX ?? [], title: hTitle };
+  // Multi-line / rotated band labels need the x-axis title pushed below the whole
+  // measured label band (max band height across panels for free scales), instead
+  // of the renderer's fixed single-line offset.
+  const TICK_CHROME_PX = 9; // tickLength + tickLabelGap defaults
+  const TITLE_GAP_PX = 10; // sits within the axis-title reserve band
```
[the remark this answers is not in the document]
Comment (ljodea) [reply packages/core/src/pipeline/assemble-scene-panels.ts]: **Fixed** in 63ffc09 — valid P3.

The band x-title offset now derives its tick chrome from the active theme, computed exactly as the SVG renderer does (`(theme.ticksX ? theme.tickLength : 0) + 3`) and threaded into `assembleScenePanels` as `tickChromePx`, instead of the hard-coded 9px. A custom theme with longer (or hidden) ticks now keeps the title clear of the wrapped/rotated label band.

File: packages/core/src/pipeline/assemble-render-model.ts [hunk 9]
```diff
@@ -16,54 +16,124 @@ import type { RenderModel } from "./types.js";
...
+} as const;
+
+function bandGuideDiagnostic(code: string, aesthetic: "x" | "y") {
+  const margin = code === "band-label-margin-overflow";
+  return {
+    code,
+    severity: "warning" as const,
+    path: `/scales/${aesthetic}`,
+    problem: margin
+      ? "A rotated categorical label is truncated to fit the axis margin cap."
```
[the remark this answers is not in the document]
Comment (ljodea) [reply packages/core/src/pipeline/assemble-render-model.ts]: **Fixed** in 63ffc09 — valid P3.

`bandGuideDiagnostic` now takes the plan mode. A `band-label-margin-overflow` on a **single-line** axis is described as a horizontal end-cap problem ("truncated to fit the axis width", cause names the panel edge / side margin, fix suggests more width), and only the rotated case mentions bottom-margin/height. No more steering users toward chart height for what is a horizontal-space problem.

Regression: `pipeline-band.test.ts` › "describes a single-line end-cap overflow in terms of width, not rotation" (asserts the problem text avoids /rotat/i and the cause/fix mentions width).

**Comment by ljodea** [comment]
@codex review

Round 2 of shepherd feedback addressed in 63ffc09 — all four P2 and both P3 findings from the latest review are fixed with regression tests:

- P2 left/right band overhang split (reserve left margin)
- P2 rotated footprint recomputed from labels that survive thinning
- P2 thinning gated on displayed tick count, not full domain
- P2 faceted band axes re-planned at the final panel size (second shared pass)
- P3 theme-aware band x-title offset
- P3 mode-aware margin-overflow diagnostic wording

Full core+spec (1644) and component (2922 + 58) suites green; check / lint / type-aware / knip / fmt all clean. Note CI is a known blocker per the issue and is being ignored for this pass.

File: packages/core/src/layout/band-guide.ts [hunk 10]
```diff
@@ -0,0 +1,443 @@
...
+      degraded: marginOverflow ? ["band-label-margin-overflow"] : [],
+    };
+  };
+
+  // Vertical band axes are not planned here (legacy path owns them).
+  if (input.orient === "vertical" || entries.length === 0) return singlePlan();
+
+  const budget = (px: number) => (quantum > 0 ? Math.floor(px / quantum) * quantum : px);
+  const bandBudget = budget(bandWidth);
+  const orthoCap = budget(orthogonalMarginCapPx);
```
[the remark this answers is not in the document]
Comment (ljodea) [reply packages/core/src/layout/band-guide.ts:282]: **Fixed** in 10aadd9 — valid P2.

The band planner now subtracts tick chrome (`tickLength + tickLabelGap`, threaded as `orthogonalChromePx`) from its orthogonal budget before accepting/truncating labels. Since `layoutPass` adds that chrome to `labelBandHeight` and the renderer offsets labels by it, the band + chrome together now honor the bottom cap — a near-cap label is truncated (with `band-label-margin-overflow`) instead of silently clipped.

Regression: `layout.test.ts` › "reserves tick chrome: band height + tickLength/gap stays within the bottom cap" (asserts `bandLabelBandHeight ≤ cap − chrome`).

File: packages/core/src/render-svg-scene.ts [hunk 11]
```diff
@@ -36,9 +36,28 @@ function renderPanelAxes(panel: ScenePanel, theme: ThemeTokens): string {
...
       if (tick.label !== "") {
-        parts.push(
-          `<text y="${px((theme.ticksX ? theme.tickLength : 0) + 3)}" dy="0.71em" text-anchor="middle" fill="${axisText}" font-size="${px(theme.axisTextSize)}" font-weight="${theme.fontWeight}">${escapeXML(tick.label)}</text>`,
-        );
+        const yOff = (theme.ticksX ? theme.tickLength : 0) + 3;
+        const font = `fill="${axisText}" font-size="${px(theme.axisTextSize)}" font-weight="${theme.fontWeight}"`;
+        if (tick.angle !== undefined && tick.angle !== 0) {
+          // Rotated band label: hang below the axis, anchored at the tick.
+          parts.push(
+            `<text transform="translate(0,${px(yOff)}) rotate(${tick.angle})" text-anchor="end" dominant-baseline="central" ${font}>${escapeXML(tick.label)}</text>`,
```
[the remark this answers is not in the document]
Comment (ljodea) [reply packages/core/src/render-svg-scene.ts:44]: **Fixed** in 10aadd9 — valid P2.

The planner now models the renderer’s actual end-anchored footprint. Rotated labels (`text-anchor="end"`) extend up-**left** of the tick, so I added `neighbourOverlapAsym` (collides the left neighbour’s right extent against the right neighbour’s left extent) and per-label left/right extents: `leftExt = w·cos + (h/2)·sin`, `rightExt = (h/2)·sin`. The overlap check, −45/−90 selection, thinning, and end-overhang all use these bounds, so a long label after a short one can no longer pass the checks and then clip the previous tick or the left edge. (This also flows into the per-side overhang from the earlier fix, so the left margin is reserved correctly.)

Regressions: `band-guide.test.ts` › "neighbourOverlapAsym: end-anchored rotated footprint" (catches a left-heavy collision a centered model misses) and "rotated overhang is left-heavy".

File: packages/core/src/scene.ts [hunk 12]
```diff
@@ -154,12 +154,19 @@ export interface SceneTick {
   /** Standalone accessible text; contextual visible labels may be shorter. */
   fullLabel: string;
   kind: "major" | "minor";
+  /** Wrapped label lines (band axis, mode "wrapped"). */
+  lines?: string[];
+  /** Rotation in degrees (band axis, mode "rotated"): -45 | -90. */
+  angle?: number;
```
[the remark this answers is not in the document]
Comment (ljodea) [reply packages/core/src/scene.ts:160]: **Fixed** in 10aadd9 — valid P2.

The interactive Svelte renderer now consumes the new scene fields. `Axis.svelte` branches to `<tspan>` rows for `tick.lines` (wrapped, centered) and a `translate(...) rotate(tick.angle)` `text-anchor="end"` label for `tick.angle` (rotated), byte-for-byte matching `render-svg-scene.ts`; `SceneView.svelte` now uses `scene.axes.x.titleOffset ?? 32` instead of the hard-coded 32px. Wrapped/rotated band labels and the pushed-down x title now render identically in `@ggsvelte/svelte` and the string renderer. Validated with the Svelte MCP autofixer (no issues).

Regressions: `packages/svelte/tests/scene/axis.test.ts` › "renders wrapped band labels as one tspan per line" and "renders rotated band labels with a rotate() transform anchored at the tick end".

File: packages/core/src/diagnostics.ts [hunk 13]
```diff
@@ -257,6 +257,12 @@ export const PIPELINE_WARNING_CATALOG = {
   "temporal-break-outside-domain": {
     summary: "One or more explicit temporal breaks were outside the trained domain and omitted.",
   },
+  "band-label-overlap": {
+    summary: "Categorical (band) axis labels overlap even after wrapping and rotation.",
+  },
+  "band-label-margin-overflow": {
+    summary: "A rotated band label is truncated (ellipsis) to fit the bounded axis margin.",
```
[the remark this answers is not in the document]
Comment (ljodea) [reply packages/core/src/diagnostics.ts]: **Fixed** in 10aadd9 — valid P3.

The `PIPELINE_WARNING_CATALOG` summary for `band-label-margin-overflow` is now mode-neutral: "A band label is truncated (ellipsis) to fit a bounded axis margin — a rotated label past the bottom cap, or a single-line end label past the side cap." This matches the mode-aware runtime diagnostic; the generated docs search index was regenerated (`docs:search:check` green).

**Comment by ljodea** [comment]
@codex review

Round 3 addressed in 10aadd9 — all three P2 and the P3 from the latest review are fixed with regression tests:

- P2 tick-chrome budget (band + chrome now honor the bottom cap, truncate instead of clip)
- P2 renderer-matched rotated bounds (end-anchored asymmetric overlap/overhang via neighbourOverlapAsym)
- P2 Svelte renderer now consumes lines / angle / titleOffset (matches the string SVG renderer)
- P3 mode-neutral margin-overflow catalog summary

Full core+spec (1648) and component (2928 + 58) suites green; check / lint / type-aware / knip / fmt / docs currency all clean. CI remains a known blocker being ignored for this pass.

File: packages/core/src/layout/band-guide.ts [hunk 14]
```diff
@@ -0,0 +1,447 @@
...
+  let rotLeft = 0;
+  let rotRight = 0;
+  for (let i = 0; i < ticks.length; i++) {
+    if (!ticks[i]!.labeled) continue;
+    const w = measurer.measureWidth(ticks[i]!.label, fontSize);
+    rotLeft = Math.max(rotLeft, leftExtOf(w, angle) - entries[i]!.center);
+    rotRight = Math.max(rotRight, rightExtOf(angle) - (extentPx - entries[i]!.center));
+  }
+  const alongOverhang = Math.max(0, Math.min(marginCapPx, rotRight));
+  const leftOverhang = Math.max(0, Math.min(marginCapPx, rotLeft));
```
[the remark this answers is not in the document]
Comment (ljodea) [reply packages/core/src/layout/band-guide.ts:458]: **Fixed** in 85c9d2b — valid P2.

The rotated branch now truncates each label to the **tighter** of the bottom-cap budget and its own left side-cap budget. An end-anchored rotated label extends left by `leftExtOf(w)`, which must fit `center + marginCapPx`; when it does not, the label is truncated (with `band-label-margin-overflow`) instead of clamping `leftOverhang` and drawing past the edge. The width-independent residual (chiefly −90°, where the left extent is ~½ line-height) is flagged when it still exceeds the side cap.

Regression: `band-guide.test.ts` › "truncates a rotated label that overflows the side cap (not just the bottom)".

File: packages/core/src/pipeline/panel-layout-facet-margins.ts [hunk 15]
```diff
@@ -22,34 +33,59 @@ export function computeFacetGridGeometry(input: FacetGridGeometryInput): FacetGr
...
+  // Band axes measure taller wrapped/rotated label bands at the NARROWER final
+  // panel width than at the optimistic `approxW`. Re-plan once at the derived
+  // panel size and reserve the (monotonically larger) margin, so bottom-row
+  // labels / the x title never draw into adjacent chrome. Gated to band facets so
+  // temporal/continuous facets keep their exact single-pass geometry.
+  if (hasBandPlan(shared)) {
+    const shared2 = computeFacetSharedMargins({
+      ...sharedInput,
+      approxW: cells.panelW,
+      approxH: cells.panelH,
```
[the remark this answers is not in the document]
Comment (ljodea) [reply packages/core/src/pipeline/panel-layout-facet-margins.ts]: **Fixed** in 85c9d2b — valid P2.

The second shared-margin pass now receives the **total cell box** (`panelW + mMax.left + mMax.right`, `panelH + mMax.top + mMax.bottom`), matching what final placement passes to `layout()`. Previously it passed `cells.panelW` (already the inner plot width), so `layout()` subtracted margins a second time and could over-escalate a panel that actually has room to wrap.

Regression: `pipeline-band.test.ts` › "does not over-escalate a faceted panel that has room to wrap" (wide two-column facet → wrapped, not rotated).

File: packages/core/src/layout/band-guide.ts [hunk 16]
```diff
@@ -0,0 +1,447 @@
...
+  const rotatedOverlaps = (angle: number, every: number) =>
+    neighbourOverlapAsym(
+      entries
+        .filter((_, i) => i % every === 0)
+        .map((e) => ({ pos: e.center, left: leftExtOf(e.width, angle), right: rightExtOf(angle) })),
+      gap,
+    );
+  // Prefer -45 when it clears neighbours and fits the cap; else -90.
+  const fits45 = !rotatedOverlaps(-45, 1) && orthoOf(labeledMaxWidth(1), -45) <= orthoCap;
+  const angle = fits45 ? -45 : -90;
```
[the remark this answers is not in the document]
Comment (ljodea) [reply packages/core/src/layout/band-guide.ts]: **Fixed** in 85c9d2b — valid P2.

The angle decision is now `rotatedOverlaps(-45, 1) ? -90 : -45` — −90° is chosen **only** for an actual −45° collision. A −45° label that merely exceeds the bottom cap is truncated within the −45° budget; escalating to −90° for that would need at least as much bottom footprint (its footprint is the full label width) and truncate harder without resolving any overlap.

Regression: `band-guide.test.ts` › "keeps −45° when labels clear neighbours but only exceed the bottom cap".

File: packages/core/src/pipeline/layout-axis-ticks.ts [hunk 17]
```diff
@@ -37,6 +37,8 @@ export function axisTicks(
       label: tick.labeled ? tick.label : "",
       fullLabel: tick.fullLabel ?? tick.label,
       kind: tick.kind ?? "major",
+      ...(tick.lines !== undefined && { lines: tick.lines }),
+      ...(tick.angle !== undefined && { angle: tick.angle }),
```
[the remark this answers is not in the document]
Comment (ljodea) [reply packages/core/src/pipeline/layout-axis-ticks.ts:41]: **Fixed** in 85c9d2b — valid P2.

`suppressProjectedLabelOverlap` now skips planner-managed band ticks (those with `angle` set, or wrapped `lines`): the band planner already fitted them, so measuring them here as centered single-line strings and blanking them under an active x projector would wrongly drop named bars. Non-planned ticks still get the projected suppression.

Regression: `pipeline-band.test.ts` › "keeps every planned band label under an active x coord projection" (`coord: { type: "transform", x: { transform: "identity", reverse: true } }`, narrow width → all rotated labels retained).

**Comment by ljodea** [comment]
@codex review

Round 4 addressed in 85c9d2b, plus `origin/main` merged (e66b783) to resolve the conflict in `assemble-render-model.ts` (kept the new non-axis guide-plan guard + band/temporal dispatch).

Four P2 fixes with regression tests:
- Rotated labels truncate to the tighter of the bottom-cap and their own left side-cap budget (no more clip-past-edge)
- −90° chosen only for an actual −45° collision; a −45° label that only exceeds the bottom cap is truncated within the −45° budget
- Facet re-plan passes the total cell box, not the inner panel width (no double margin subtraction / over-escalation)
- suppressProjectedLabelOverlap skips planner-managed rotated/wrapped band ticks (no dropped bars under an active x projector)

Core+spec (1690) green; lint / type-aware / knip / fmt / docs-currency clean; merge resolved (MERGEABLE). CI remains a known blocker being ignored for this pass.

**Comment by ljodea** [comment]
## CI investigation — the 3 failing jobs

**1. `component-journeys (docs a11y playwright)` — pre-existing, out of scope.**
6 interaction-example a11y tests fail because `interaction-accessibility.spec.ts` injects axe-core via `page.addScriptTag({ content: axe.source })`, which the docs Content-Security-Policy (added in #391/#403) blocks as an inline script. This is **not** caused by #387: `apps/docs/csp.ts` is byte-identical to `main`, and `main`'s own CI fails the same job (runs for ff4ad4c and 557c080 both list `component-journeys` as failed). Already tracked by **#417**. The interaction examples don't involve categorical axes.

**2. `build docs, screenshot, upload candidate baselines` — expected visual change, in scope.**
Two VR diffs: `col-basic-light` (4833px / 2%) and `boxplot-by-category-light` (953px / 1%) — both **categorical-x charts** whose x-axis label layout/margins #387 intentionally changes (all band axes now go through the measured planner). This is the normal source-first VR flow: baselines regenerate **after merge** via `/approve-visuals` (renders at the merged commit in the pinned container, then lands through a human-reviewed `vr-update/pr-408` PR). `main` does not fail this job, confirming the diff is #387's own rendering change.

**3. `ci-gate (required aggregator)` — red solely because of (1) and (2).**

All four rounds of Codex review feedback are addressed (0 unanswered threads). Merging; baselines will be regenerated post-merge via `/approve-visuals`.
